### PR TITLE
feat: add Kimi-K3 model definition and quantization example

### PR DESCRIPTION
**Pull request by kylesayrs**
Currently dequantizes the entire model up-front, but we can avoid upfront quantization using [layerwise decompression](https://github.com/vllm-project/llm-compressor/pull/2995)

## Changes ##
* Add modeling code for kimi-k3 (with documentation of diff)
* Add example script using small model for initial runs
* Modify `validate_model` to be able to pass `trust_remote_code`
* Add `_apply_attn_res` to trace ignore list
* Set `add_labels=False` for prebaked datasets (do not add labels by default, this was left over from sparseml)

## Testing ##
* Quantized Kimi-K3 using this example
* Ran small model using this example

### CODE REVIEW

File: src/llmcompressor/modeling/kimi_k3/modeling_kimi_linear.py
Comment (brian-dellabetta) [inline src/llmcompressor/modeling/kimi_k3/modeling_kimi_linear.py:1]: is this needed for some reason? or just following convention in transformers src?

File: src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py
Comment (brian-dellabetta) [inline src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py:1]: do you have a record at all of what you had to change to get this to work? In case we need to modify later?

**Review by brian-dellabetta — COMMENTED** [review]
overall lgtm, couple questions. when you have empty checkboxes in your PR summary, does that mean PR is still WIP or that you've added each of those?

File: src/llmcompressor/modeling/kimi_k3/modeling_kimi_linear.py
Comment (kylesayrs) [reply src/llmcompressor/modeling/kimi_k3/modeling_kimi_linear.py:1]: Ah let me change the naming

File: src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py
Comment (kylesayrs) [reply src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py:1]: Added a diff record!

Comment (brian-dellabetta) [reply src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py:1]: sorry, where is the diff? is it a commit?

File: examples/quantizing_moe/kimi_k3_example.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+from compressed_tensors.quantization import QuantizationConfig
+from transformers import AutoTokenizer
+
+from datasets import load_dataset
+from llmcompressor import oneshot
+from llmcompressor.modeling.kimi_k3 import KimiK3ForConditionalGeneration
+from llmcompressor.modifiers.quantization import QuantizationModifier
+from llmcompressor.utils import load_context
+
+MODEL_ID = "moonshotai/Kimi-K3"
```
Comment (dsikka) [inline examples/quantizing_moe/kimi_k3_example.py]: can we include in key models: https://docs.vllm.ai/projects/llm-compressor/en/latest/key-models/

File: src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py
Comment (kylesayrs) [reply src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py:1]: There is no diff for this file, but other files have these comments
https://github.com/vllm-project/llm-compressor/pull/2994/changes#diff-31d1b29fad82afccfeab91bce7bbcb974598b53cff9089a22ee1d0dbcac01220R23

Comment (brian-dellabetta) [reply src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py:1]: Hmm, it's not loading correctly for me. can you point to the file name / line number? This is also for personal edification, but a README outlining changes from what was copy-pasted from the model checkpoint itself would be handy

File: docs/key-models/kimi-k3/nvfp4-example.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+from compressed_tensors.distributed import init_dist
+from transformers import AutoConfig, AutoProcessor, CompressedTensorsConfig
+
+from llmcompressor import oneshot
+from llmcompressor.modeling.kimi_k3 import KimiK3ForConditionalGeneration
+from llmcompressor.modifiers.quantization import QuantizationModifier
+from llmcompressor.utils import load_context
+
+MODEL_ID = "inference-optimization/Kimi-K3-0.40B-MXFP4"  # "moonshotai/Kimi-K3"
+
```
Comment (dsikka) [inline docs/key-models/kimi-k3/nvfp4-example.md:24]: should be the original?

Comment (dsikka) [reply docs/key-models/kimi-k3/nvfp4-example.md:24]: are we dequantizing first?

File: examples/quantizing_moe/kimi_k3_example.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from compressed_tensors.distributed import init_dist
+from transformers import AutoConfig, AutoProcessor, CompressedTensorsConfig
+
+from llmcompressor import oneshot
+from llmcompressor.modeling.kimi_k3 import KimiK3ForConditionalGeneration
+from llmcompressor.modifiers.quantization import QuantizationModifier
+from llmcompressor.utils import load_context
+
+MODEL_ID = "inference-optimization/Kimi-K3-0.40B-MXFP4"  # "moonshotai/Kimi-K3"
+
```
Comment (dsikka) [inline examples/quantizing_moe/kimi_k3_example.py:12]: same comments as above

File: src/llmcompressor/entrypoints/oneshot.py [hunk 4]
```diff
@@ -362,6 +364,7 @@ def oneshot(
         "_prepare_4d_causal_attention_mask_with_cache_position",
         "_update_linear_attn_mask",
         "project_per_layer_inputs",
+        "_apply_attn_res",
```
Comment (dsikka) [inline src/llmcompressor/entrypoints/oneshot.py:367]: why is this needed?

File: src/llmcompressor/transformers/data/base.py [hunk 5]
```diff
@@ -128,7 +128,7 @@ def __init__(
             self.max_seq_length = None
             self.padding = False
 
-    def __call__(self, add_labels: bool = True) -> DatasetType:
+    def __call__(self, add_labels: bool = False) -> DatasetType:
```
Comment (dsikka) [inline src/llmcompressor/transformers/data/base.py:131]: why did we change the default

**Review by dsikka — COMMENTED** [review]
Initial comments

File: docs/key-models/kimi-k3/nvfp4-example.md (hunk 2 above)
Comment (kylesayrs) [reply docs/key-models/kimi-k3/nvfp4-example.md:24]: Yes https://github.com/vllm-project/llm-compressor/pull/2994/changes#diff-fccc0621a17b600cfbe5c01f57e62cf67cbb8d39a95d62b2b3d3e9f9bfc70b08R24-R28

Comment (kylesayrs) [reply docs/key-models/kimi-k3/nvfp4-example.md:24]: I meant to use the small model. Imho a user should always do the small model first before swapping to the large checkpoint

File: src/llmcompressor/entrypoints/oneshot.py (hunk 4 above)
Comment (kylesayrs) [reply src/llmcompressor/entrypoints/oneshot.py:367]: Enables tracing at the (attention + experts) granularity

File: src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py
Comment (kylesayrs) [reply src/llmcompressor/modeling/kimi_k3/configuration_kimi_k3.py:1]: I've confirmed that this is loading correctly for me

File: src/llmcompressor/transformers/data/base.py (hunk 5 above)
Comment (kylesayrs) [reply src/llmcompressor/transformers/data/base.py:131]: This model has wonky logic around computing loss and gradients. It only does it if labels are provided. It's better to skip for this model, and for models in general

File: docs/key-models/kimi-k3/nvfp4-example.md (hunk 2 above)
Comment (dsikka) [reply docs/key-models/kimi-k3/nvfp4-example.md:24]: Ok we should move this model under the CI modes list or something so it doesn't get deleted. I would also mention how this is first decompressed

**Review by dsikka — COMMENTED** [review]
Will approve after the comment is addressed

**Comment by kylesayrs** [comment]
@dsikka comments addressed, model is in HF CI collection

**Review by dsikka — APPROVED** [review]
